makeflow/src/mountfile.c

```diff
@@ -0,0 +1,660 @@
+			LDEBUG("mount_check(%s, %s) failed: %s!\n", source, target, strerror(errno));
+			err_num++;
+			continue;
+		}
+
+		p = xxstrdup(source);
+		if(!p) {
+			LDEBUG("xxstrdup(%s) failed!\n", source);
+			err_num++;
+		}
+
+		/* df->source may already be set based on the information from the makeflow log file, so free it first. */
+		if(df->source) free(df->source);
+
+		df->source = p;
+	}
+
+	if(fclose(f)) {
+		LDEBUG("fclose(`%s`) failed: %s!\n", mountfile, strerror(errno));
+		return -1;
+	}
+
+	if(err_num) return -1;
+
+	return 0;
+}
+
+/* check_cache_dir checks the validity of the cache dir, and create it if possible.
+ * @param cache: a file path.
+ * return 0 on success, return non-zero on failure.
+ */
+int check_cache_dir(const char *cache) {
+	struct stat st;
+
+	/* Checking principle: the cache must locate under the CWD. */
+	if(!cache || !*cache) {
+		LDEBUG("the cache (%s) can not be empty!\n", cache);
+		fprintf(stderr, "the cache (%s) can not be empty!\n", cache);
+		return -1;
+	}
+
+	/* Check whether the cache is an absolute path. */
+	if(cache[0] == '/') {
+		LDEBUG("the cache (%s) should not be an absolute path!\n", cache);
+		fprintf(stderr, "the cache (%s) should not be an absolute path!\n", cache);
+		return -1;
+	}
+
+	/* check whether cache includes .. */
+	if(path_has_doubledots(cache)) {
+		LDEBUG("the cache (%s) include ..!\n", cache);
+		fprintf(stderr, "the cache (%s) include ..!\n", cache);
+		return -1;
+	}
+
+	/* check whether cache includes any symlink link, this check prevent the makeflow breaks out the CWD. */
+	if(path_has_symlink(cache)) {
+		LDEBUG("the cache (%s) should not include any symbolic link!\n", cache);
+		fprintf(stderr, "the cache (%s) should not include any symbolic link!\n", cache);
+		return -1;
+	}
+
+	/* Check whether cache already exists. */
+	if(access(cache, F_OK)) {
+		mode_t default_dirmode = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
+		if(errno != ENOENT) {
+			LDEBUG("access(%s) failed: %s\n", cache, strerror(errno));
+			return -1;
+		}
+
+		LDEBUG("the cache (%s) does not exist, creating it ...\n", cache);
+		if(mkdir_recursive(cache, default_dirmode)) {
+			LDEBUG("mkdir_recursive(%s) failed: %s\n", cache, strerror(errno));
+			return -1;
+		}
+		return 0;
+	}
+
+	/* check whether cache is a dir */
+	if(lstat(cache, &st) == 0) {
+		if(!S_ISDIR(st.st_mode)) {
+			LDEBUG("the cache (%s) should be a dir!\n", cache);
+			fprintf(stderr, "the cache (%s) should be a dir!\n", cache);
+			return -1;
+		}
+	} else {
+		LDEBUG("lstat(%s) failed: %s!\n", cache, strerror(errno));
+		fprintf(stderr, "lstat(%s) failed: %s!\n", cache, strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int mount_install_all(struct dag *d) {
+	struct list *list;
+	struct dag_file *df;
+	char *cache_dir;
+
+	if(!d) return 0;
+
+	if(d->cache_dir) {
+		/* check the validity of the cache_dir and create it if neccessary and feasible. */
+		if(check_cache_dir(d->cache_dir)) {
```

Comment:
I understand, but if the function works for both cases, even if it is overkill, then I think we can use it for both. This would allow us to log the creation in check_cache_dir and just set the cache_dir name if it were null.